### PR TITLE
[SPARK-19923][SQL] Remove unnecessary type conversions per call in Hive

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr removed unnecessary type conversions per call in Hive: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala#L116

## How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by maropu** [comment]
I'm not sure this makes sense, so could you check? cc: @hvanhovell

**Comment by maropu** [comment]
I quickly did check performance changes;
```
public class TestGenericUDF extends GenericUDF {
  @Override
  public ObjectInspector initialize(ObjectInspector[] objectInspectors) throws UDFArgumentException {
      return PrimitiveObjectInspectorFactory.javaLongObjectInspector;
  }

  @Override
  public Object evaluate(DeferredObject[] args) throws HiveException {
      final long a1 = PrimitiveObjectInspectorFactory.javaLongObjectInspector.get(args[0].get());
      final long a2 = PrimitiveObjectInspectorFactory.javaLongObjectInspector.get(args[1].get());
      final long a3 = PrimitiveObjectInspectorFactory.javaLongObjectInspector.get(args[2].get());
      final long a4 = PrimitiveObjectInspectorFactory.javaLongObjectInspector.get(args[3].get());
      return a1 + a2 + a3 + a4;
  }
}
```

```
$./bin/spark-shell --master local[1]  --conf spark.sql.shuffle.partitions=1 -v

scala> sql("CREATE TEMPORARY FUNCTION testUdf AS 'hivemall.ftvec.TestGenericUDF'")
scala> :paste
def timer[R](block: => R): R = {
  val t0 = System.nanoTime()
  val result = block
  val t1 = System.nanoTime()
  println("Elapsed time: " + ((t1 - t0 + 0.0) / 1000000000.0)+ "s")
  result
}

scala> spark.range(3000000).createOrReplaceTempView("t")
scala> timer { sql("SELECT testUdf(id, id, id, id) FROM t").queryExecution.executedPlan.execute().foreach(x => {}) }
```

```
# performance w/ this patch
Elapsed time: 1.901269167s

# performance w/o this patch
Elapsed time: 0.492860666s
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 1]
```diff
@@ -208,10 +215,8 @@ private[orc] class OrcSerializer(dataSchema: StructType, conf: Configuration)
       oi.setStructFieldData(
         struct,
         fieldRefs.get(i),
-        wrap(
-          row.get(i, dataSchema(i).dataType),
-          fieldRefs.get(i).getFieldObjectInspector,
-          dataSchema(i).dataType))
+        wrappers(i)(row.get(i, dataSchema(i).dataType))
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Hi, @maropu .
This seems to be improvement, too. Could you add some description about this?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -196,6 +198,11 @@ private[orc] class OrcSerializer(dataSchema: StructType, conf: Configuration)
 
   private[this] val cachedOrcStruct = structOI.create().asInstanceOf[OrcStruct]
 
+  private[this] val wrappers = dataSchema.zip(structOI.getAllStructFieldRefs().asScala.toSeq)
+      .map { case (f, i) =>
+    wrapperFor(i.getFieldObjectInspector, f.dataType)
+  }
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: (my personal taste that might be worth considering.. 

```scala
private[this] val wrappers = dataSchema
  .zip(structOI.getAllStructFieldRefs.asScala.toSeq)
  .map { case (f, i) => wrapperFor(i.getFieldObjectInspector, f.dataType) }
```

or maybe

```scala
private[this] val wrappers = 
  structOI.getAllStructFieldRefs.asScala.toSeq.zip(dataSchema).map { case (ref, field) =>
    wrapperFor(ref.getFieldObjectInspector, field.dataType)
  }
```

which resembles https://github.com/apache/spark/blob/226d38840c8d3f40639715d755df6fb4fee2715f/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala#L364-L366
)

**Comment by maropu** [comment]
Thanks for valuable comments! I'll update soon.

**Comment by hvanhovell** [comment]
@maropu this looks like a nice improvement. It is an idea to check all uses of `wrap`, and see if we can replace them by `wrapperFor`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala (hunk 1 above)
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Added the description to explain `wrappers`: https://github.com/apache/spark/pull/17264/files#diff-01999ccbf13e95a0ea2d223f69d8ae23R201

**Comment by maropu** [comment]
@hvanhovell I checked all the related code and then I think this pr includes all the place we could replace `wrap` with `wrapperFor`.

**Comment by hvanhovell** [comment]
LGTM merging to master. Thanks!

**Comment by rxin** [comment]
In the future can we put the perf result in PR descriptions?

**Comment by maropu** [comment]
@rxin okay!

**Pull request closed without merging**